include/merlin/memory_pool.hpp

```diff
@@ -0,0 +1,591 @@
+};
+
+/**
+ * Forward declares required to make templated ostream overload work.
+ */
+template <class Allocator>
+class MemoryPool;
+
+template <class Allocator>
+std::ostream& operator<<(std::ostream&, const MemoryPool<Allocator>&);
+
+/**
+ * CUDA deferred execution aware memory pool implementation. As for every memory
+ * pool, the general idea is to have resuable buffers. All buffers have the same
+ * size.
+ *
+ * General behavior:
+ *
+ * This memory pool implementation attempts to avoid blocking before the fact,
+ * but also avoids relying on a background worker.
+ *
+ * Buffer borrow and return semantics tightly align with C++ RAII principles.
+ * That is, if a workspace is requested, any borrowed buffers will be returned
+ * automatically when leaving the scope.
+ *
+ * You can either borrow a single buffer, or a workspace (that is multiple
+ * buffers). We support dynamic and static workspaces. Static workspaces have
+ * the benefit that they will never require heap memory (no hidden allocations).
+ *
+ *
+ * Buffer borrowing:
+ *
+ * If buffers are requested, we take them from the stock, if available. If the
+ * stock is depleted, we check if any pending buffer has been used up by the GPU
+ * and adds them to the stock. If was also not successful, we allocate a new
+ * buffer. Buffers or workspaces (groups of buffers).
+ *
+ * When borrowing a buffer a streaming context can be specified. This context is
+ * relevant for allocation and during returns. It is assumed that the stream you
+ * provide as context will be the stream where you queue the workload. Not doing
+ * so may lead to undefined behavior.
+ *
+ * Buffer return:
+ *
+ * If no context is provided, we cannot make any assumptions regarding the usage
+ * one the device. So we sychronize the device first and then return the buffer
+ * to the stock. If a streaming context was provided, we queue an event and add
+ * the buffer to the `pending` pool. That means, the buffer has been
+ * reqlinquished by the CPU, but may still be used by the GPU. If no pending
+ * slot is available, we probe the currently pending buffers events for
+ * completion. Completed pending buffers are returned to the reserve. If so, we
+ * queue the buffer in the freed slot. If that was unsucessful (i.e., all
+ * currently pending buffers are still in use by the GPU), we have no choice but
+ * the free the buffer using the current stream.
+ *
+ * In either case, `max_reserve` represents the maxmum size of the stock. If
+ * returning a buffer would lead to the stock exeeding this quantity, the buffer
+ * is queued for destruction.
+ */
+template <class Allocator>
+class MemoryPool final {
+ public:
+  using pool_type = MemoryPool<Allocator>;
+  using alloc_type = typename Allocator::type;
+  template <class Container>
+  class Workspace {
+   public:
+    inline Workspace() : pool_{nullptr}, stream_{0} {}
+
+    inline Workspace(pool_type* pool, cudaStream_t stream)
```

Comment:
Possibly, can give a default value 0 to the stream, like `stream=0`.

Reply:
It is internal API. Arguably, doesn't need a default argument. I just pushed an update that makes these constructors private.

Reply:
Please resolve if acceptable.